### PR TITLE
ttnn.concat fix for rank one tensors

### PR DESCRIPTION
**Pull request by jbedichekTT**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/22541)

### Problem description
Rank 1 tensors passed to `ttnn.concat` causes errors.  

### What's changed
Added unsqueezing step.

The error being thrown was in the slice op, line 80:
```
TT_FATAL(
        input_tensor_a.get_padded_shape().rank() == this->slice_start.rank() &&
            this->slice_start.rank() == this->slice_end.rank(),
        "Error");
```
The slice op is called inside the `build_untilize_rm_retilize_concat` function, in order to slice away the padding for concatenating the logical parts of the tensor.  The relevant tensors attributes that were triggering the error were the rank of the input tensor not matching the rank of the slice start and/or end.

Unsqueezing was a quick fix to make the ranks always equal, but it could be glossing over a deeper issue in how those values are constructed, in this part of the code (In the .pre_transform of `build_untilize_rm_retilize_concat`):

```auto untilized_tensor = ttnn::untilize(input_tensor);
                    // untilized, so now we have a padded rm tensor. we slice to
                    // remove the padding.
                    const auto& input_shape = input_tensor.get_logical_shape();
                    std::vector<uint32_t> begins_vec(input_shape.rank(), 0);
                    tt::stl::Span<const uint32_t> begins = begins_vec;
                    tt::stl::Span<const uint32_t> ends = input_shape.view();
                    std::vector<uint32_t> steps_vec(input_shape.rank(), 1);
                    tt::stl::Span<const uint32_t> steps = steps_vec;
```

I can't find anything wrong with this, but `begins` and `ends` translate to inconsistent ranks when the input is 1D.  The unsqueezing step seems to fix this, but is there a better solution?

### Second Bug 
There was also a second unrelated error that came up that I added a fix for in this PR, see this GitHub issue: [link](https://github.com/tenstorrent/tt-metal/issues/22845)

This occurred because the original implementation gave every core metadata for all input tensors, resulting in `3 + 5×N `runtime arguments per core for N tensors, which exceeds the hardware limit of 256 arguments when trying to concatenate 51 tensors or more at one time.

To fix this, the logic which distributes core workloads was modified to pre-compute which data each core needs before reading the tensors, as opposed to figuring it out at runtime on each core.  

This should improve memory usage by eliminating redundant tensor meta data being passed to multiple cores.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp [hunk 1]
```diff
@@ -64,7 +62,20 @@ MassagedConcat build_unsqueeze_concat(int input_rank, const MemoryConfig& output
                 tensors.end(),
                 std::back_inserter(itensors),
                 [](const ttnn::Tensor& input_tensor) -> ttnn::Tensor { return ttnn::unsqueeze_to_4D(input_tensor); });
-            return std::make_tuple(itensors, dim + 4 - input_rank, groups);
+
+            int adjusted_dim;
+            if (dim < 0) {
+                adjusted_dim = 4 + dim;
+            } else {
+                adjusted_dim = dim + (4 - input_rank);
```
[the remark this answers is not in the document]
Comment (jbedichekTT) [reply ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp:70]: I added in an unsqueezing step, so It's being unsqueezed to reach a rank of 4,  and the extra dimensions get appended to the beginning of the tensor, so what was the 0th dimension is now the 1st dimension.   

i.e
shape: `[2, 2, 2]` catting to `[2, 2, 2]` in the 0th dimension becomes:
`[1, 2, 2, 2]` and `[1, 2, 2, 2]` after unsqueezing, then
`[1, 4, 2, 2]` where the adjusted dim of 1 was catted.

Was the unsqueezing step intentionally not used though? is it worse for performance?

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp [hunk 2]
```diff
@@ -268,7 +279,7 @@ ttnn::Tensor ConcatOperation::invoke(
     const ttnn::Tensor& first_tensor = input_tensors.front();
     const int rank = first_tensor.get_logical_shape().rank();
 
-    dim = first_tensor.get_padded_shape().get_normalized_index(dim);
+    dim = first_tensor.get_logical_shape().get_normalized_index(dim);
```
Comment (jbedichekTT) [inline ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp:282]: I'm not sure about this, does anyone know?

**Review by llongTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp (hunk 2 above)
Comment (llongTT) [reply ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp:282]: The logical shape is the shape of your actual data.
Padded shape is usually a bit larger due to padding to certain dimension, like tile_layout need to have tile size 32x32, if your height or width is less than 32, it will be padded to 32.

File: open_mpi.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+    echo "→ Downloading $DEB_FILE …"
+    wget -q --show-progress -O "$TMP_DIR/$DEB_FILE" "$DEB_URL"
+
+    # 2. Install
+    echo "→ Installing $DEB_FILE …"
+    apt-get update -qq
+    apt-get install -f -y "$TMP_DIR/$DEB_FILE"
+}
+install_mpi_uflm
```
Comment (philei-tt) [inline open_mpi.sh:18]: What's that? Should it be removed?

File: test.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
+import torch
```
Comment (philei-tt) [inline test.py:1]: You can't just push some random python file into root of the project. Can you please make pytests for those broken cases? You should probably `update/extend tests/ttnn/unit_tests/operations/test_concat.py`

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp [hunk 5]
```diff
@@ -64,7 +62,20 @@ MassagedConcat build_unsqueeze_concat(int input_rank, const MemoryConfig& output
                 tensors.end(),
                 std::back_inserter(itensors),
                 [](const ttnn::Tensor& input_tensor) -> ttnn::Tensor { return ttnn::unsqueeze_to_4D(input_tensor); });
-            return std::make_tuple(itensors, dim + 4 - input_rank, groups);
+
+            int adjusted_dim;
+            if (dim < 0) {
```
Comment (philei-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp:67]: Does it make negative dim work? If so, please mention this in the PR description that you fixed that, and also add test case for negative dimensions

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp [hunk 6]
```diff
@@ -64,7 +62,20 @@ MassagedConcat build_unsqueeze_concat(int input_rank, const MemoryConfig& output
...
+                adjusted_dim = 4 + dim;
+            } else {
+                adjusted_dim = dim + (4 - input_rank);
+            }
+
+            concat_db_print(
```
Comment (philei-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp:73]: `concat_db_print ` shouldn't exist. Please use `tt::log_debug` for such things, or just remove if you finished debugging your code

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp (hunk 5 above)
Comment (philei-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp:67]: Also, what it 4? Magic number is not good, either give it a proper name or fix this if this is not generalized solution.

File: ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp [hunk 7]
```diff
@@ -720,16 +720,38 @@ tt_metal::operation::ProgramWithCallbacks sharded_concat_multi_core(
     }
 }
 
+// Add this include at the top of your file:
```
Comment (philei-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp:723]: Are those copilot's comment?) ahaha remove them please

File: ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp [hunk 8]
```diff
@@ -720,16 +720,38 @@ tt_metal::operation::ProgramWithCallbacks sharded_concat_multi_core(
...
+#include <tt-metalium/distributed.hpp>
+
+// Fixed version of your concat_multi_core function:
```
Comment (philei-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp:726]: 👀

File: ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp [hunk 9]
```diff
@@ -720,16 +720,38 @@ tt_metal::operation::ProgramWithCallbacks sharded_concat_multi_core(
...
+
 tt_metal::operation::ProgramWithCallbacks concat_multi_core(
     const std::vector<Tensor>& input_tensors, const uint32_t dim, const Tensor& output) {
     tt_metal::Program program = tt_metal::CreateProgram();
 
-    tt_metal::IDevice* device = output.device();
+    // FIXED: Cast to MeshDevice pointer to access mesh methods
+    tt_metal::IDevice* raw_device = output.device();
+    auto mesh_device = dynamic_cast<tt_metal::distributed::MeshDevice*>(raw_device);
+    TT_FATAL(mesh_device != nullptr, "Expected MeshDevice but got different device type");
```
Comment (philei-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp:751]: I think you can just do `auto* mesh_device = output.mesh_device()`

File: ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp [hunk 10]
```diff
@@ -878,80 +882,102 @@ tt_metal::operation::ProgramWithCallbacks concat_multi_core(
...
+            // Read current metadata using NEW API
+            std::vector<uint32_t> metadata_data;
+            metadata_data.resize((metadata_buffer->size() + sizeof(uint32_t) - 1) / sizeof(uint32_t));
+            tt_metal::distributed::MeshCoordinate mesh_coord{0, 0};
+            tt_metal::distributed::ReadShard(
+                mesh_device->mesh_command_queue(0), metadata_data, metadata_buffer, mesh_coord);
+
+            // Update tensor addresses
+            for (uint32_t i = 0; i < num_input_tensors; ++i) {
+                uint32_t addr_index = 4 + i * 5;
```
Comment (philei-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp:963]: magic numbers?

**Review by philei-tt — CHANGES_REQUESTED** [review]
Please remove files from root of project, and add proper tests for all the broken tests. Also, address other comments

Also, I can't infer your solution from the code. Can you clearly describe what is done in PR message? Are you squeezing, concating and then unsqueezing? Please ping once done so that I can review again

**Review by ntarafdar — APPROVED** [review]

**Comment by mmanzoorTT** [comment]
@jbedichekTT any ETA for this PR?

File: open_mpi.sh (hunk 3 above)
Comment (blozano-tt) [reply open_mpi.sh:18]: Looks like this was extracted from install.sh ?

File: test.py (hunk 4 above)
Comment (blozano-tt) [reply test.py:1]: Agree.

**Review by blozano-tt — CHANGES_REQUESTED** [review]
